cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -287,233 +319,217 @@ class TypedConverter : public SeqConverter {
   BuilderType* typed_builder_;
 };
 
-class BoolConverter : public TypedConverter<BooleanBuilder> {
+template <typename BuilderType>
+class TypedConverterVisitor : public TypedConverter<BuilderType> {
  public:
-  Status AppendData(PyObject* seq) override {
-    int64_t size = static_cast<int64_t>(PySequence_Size(seq));
-    RETURN_NOT_OK(typed_builder_->Reserve(size));
-    for (int64_t i = 0; i < size; ++i) {
-      OwnedRef item(PySequence_GetItem(seq, i));
-      if (item.obj() == Py_None) {
-        typed_builder_->AppendNull();
-      } else {
-        if (item.obj() == Py_True) {
-          typed_builder_->Append(true);
-        } else {
-          typed_builder_->Append(false);
-        }
+  Status AppendData(PyObject* obj, int64_t size) override {
+    /// Ensure we've allocated enough space
+    RETURN_NOT_OK(this->typed_builder_->Reserve(size));
+    // Iterate over the items adding each one
+    if (PySequence_Check(obj)) {
+      for (int64_t i = 0; i < size; ++i) {
+	OwnedRef ref(PySequence_GetItem(obj, i));
+	RETURN_NOT_OK(appendItem(ref));
+      }
+    } else if (PyObject_HasAttrString(obj, "__iter__")) {
+      PyObject* iter = PyObject_GetIter(obj);
+      PyObject* item;
+      while ((item = PyIter_Next(iter))) {
+	OwnedRef ref(item);
+	RETURN_NOT_OK(appendItem(ref));
       }
+      Py_DECREF(iter);
+    } else {
+      return Status::TypeError("Object is not a sequence or iterable");
     }
     return Status::OK();
   }
+
+  virtual Status appendItem(OwnedRef &item) = 0;
 };
 
-class Int64Converter : public TypedConverter<Int64Builder> {
+class BoolConverter : public TypedConverterVisitor<BooleanBuilder> {
  public:
-  Status AppendData(PyObject* seq) override {
-    int64_t val;
-    int64_t size = static_cast<int64_t>(PySequence_Size(seq));
-    RETURN_NOT_OK(typed_builder_->Reserve(size));
-    for (int64_t i = 0; i < size; ++i) {
-      OwnedRef item(PySequence_GetItem(seq, i));
-      if (item.obj() == Py_None) {
-        typed_builder_->AppendNull();
+  inline Status appendItem(OwnedRef &item) final {
```

Comment:
I don't think `inline` marker is needed. We could go to some effort using CRTP to inline this method (https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern), in which case we wouldn't have any virtual method